File: src/main/java/leetcode/hot100/LeetCode287.java
package leetcode.hot100;

/**
 * @author wujingxinit@outlook.com
 * @date 2022/9/14 01:44
 */
public class LeetCode287 {

    public int findDuplicate(int[] nums) {

        if (nums == null || nums.length == 0) {
            return -1;
        }

        int left = 0;
        int right = nums.length - 1;
        int ans = -1;
        while (left < right) {
            int mid = (left + right) / 2;
            int cnt = 0;
            for (int num : nums) {
                if (num <= mid) {
                    cnt++;
                }
            }
            if (cnt <= mid) {
                left = mid + 1;
            } else {
                right = mid - 1;
                ans = mid;
            }
        }
        return ans;
    }
}

File: go/src/leetcode/hot100/LeetCode287.go
package hot100

func findDuplicate(nums []int) int {
	if len(nums) == 0 {
		return 0
	}
	left, right, ans := 0, len(nums)-1, -1
	for left <= right {
		mid := (left + right) / 2
		count := 0
		for i := 0; i < len(nums); i++ {
			if nums[i] <= mid {
				count++
			}
		}
		if count <= mid {
			left = mid + 1
		} else {
			right = mid - 1
			ans = mid //说明这是符合条件的答案， 为什么不直接返回呢？
			//因为要找到第一个数字，那个数字才是答案
		}
	}
	return ans
}

func findDuplicateII(nums []int) int {
	slow, fast := 0, 0
	for slow, fast = nums[slow], nums[nums[fast]]; slow != fast; slow, fast = nums[slow], nums[nums[fast]] {
	}
	slow = 0
	for slow != fast {
		slow = nums[slow]
		fast = nums[fast]
	}
	return slow
}
